### PR TITLE
Unarmed_weapon behaviour alterations

### PR DESCRIPTION
**Pull request by Eighth-of-Third**
Non-bitrotten version of #18543. Adds unarmed_weapon to all tonfas.

Splits disarm immunity into a separate flag, and adds it to all previously existing unarmed weapons. This allows for disarmable unarmed weapons, and potentially disarm immune regular weapons. Tonfas are disarmable.

Allows unarmed weapons to be non-conductive for the purposes of zapback based on their materials, like other weapons. All previously existing unarmed weapons are conductive; currently wooden and plastic tonfas are the only non-conductive unarmed weapons.

Removes redundant weapon entries and one technique from the martial arts jsons. Biojutsu allows the use of zero non-unarmed weapons, so there's no reason for the biojutsu_rapid_armed technique to exist.

Fixes Silat techniques not being unarmed_allowed. The style itself is unarmed_allowed, but none of the techniques were, which meant the style didn't actually support the player being unarmed.

### CODE REVIEW

**Comment by DangerNoodle** [comment]
I would suggest adding `NO_DISARM` to the bionic claws and monomolecular blade. If `NO_UNWIELD` already precludes disarming, then it was a pre-existing redundancy that is likely harmless to retain. If disarm immunity for bioweapons was controlled solely through `UNARMED_WEAPON` previously, then the new flag is now necessary.

**Comment by Izicata** [comment]
I'm one step ahead of you: is_disarmable checks for both no_disarm and no_unwield.

**Comment by DangerNoodle** [comment]
Good to have ensured that the code already accounted for this. As I said, the redundancy should be harmless, but glad to see it was not required.

**Comment by mugling** [comment]
I agree with @Coolthulu's comments in the previous PR. You need to separate out the tonfa from the other changes.

**Comment by mugling** [comment]
Also `NO_DISARM` is overly redundant - why can't we just check for `NO_UNWIELD`?

**Comment by Zireael07** [comment]
> Also NO_DISARM is overly redundant - why can't we just check for NO_UNWIELD?

I don't think unarmed weapons are NO_UNWIELD. Also the OP said:
> This allows for disarmable unarmed weapons, and potentially disarm immune regular weapons.

... so we could have weapons that the player can unwield but that are disarm immune. 
Example: Medieval knights had those fancy platemail gloves that had hooks or something to prevent being disarmed. So they could unwield the weapon themselves (take off the glove and therefore the weapon) but they couldn't be disarmed.

**Comment by mugling** [comment]
Multiple flags adds a lot of code complexity. A crude metric is every line you write will be read 100 times. That's a lot of peoples time for a really minor detail like that.

**Comment by Izicata** [comment]
>I agree with @Coolthulu's comments in the previous PR. You need to separate out the tonfa from the other changes.

If you agree with his comments, are you going to respond to my arguments explaining why his comments are incorrect, or are you just going to stonewall?

>A tonfa boosting kicks is no weirder than brass knuckles boosting kicks; allowing brass knuckles to work with martial arts and not tonfas is inconsistent. Either tonfas should have the unarmed_weapon flag or all unarmed weapons should only be allowed for specific martial arts, and I'm not going to go through all the unarmed weapons manually adding them to the weapons lists of unarmed_allowed styles. The whole reason the unarmed_weapon flag exists is to avoid having to do that.

>And again, I always thought kicks benefiting from unarmed_allowed weapons was simulating using the weapon before or after the kick; i.e. a punch, then a kick, as part of the same attack. Or a kick and followup punch. Especially in taekwondo; a rough translation of the name would be "way of the foot and fist". It's not just kicks. Combination attacks mixing both kicks and punches are a staple of the style.

**Comment by Coolthulhu** [comment]
> This will be the third time I've posted this argument, and I have not received any opposing argument besides "I feel that it's stupid and weird".

OK, here are some arguments:
* Just because the game has faults doesn't mean we will accept all shit that makes them worse.
* It's your job to present an argument for why will it improve the game, not our to tell you why not, at least until you present a valid case for your side.
* Tonfa is "less unarmed" than brass knuckles. It won't support simple jabs as well as brass knuckles and requires different movement to utilize fully. Boxer won't utilize tonfa well, but could do fine with brass knuckles.
* Someone not trained in unarmed combat will utilize tonfa as a club.
* Tonfa's stats are for being used as a club, not as brass knuckles. For example, relatively high damage that requires momentum. Brass knuckles are harder than a tonfa, but hit weaker because they do not get that momentum
* Your argument, that we won't be making things worse, is incorrect. Tonfa is closer to held rock than to brass knuckles.

I see the following options:
* Unarmed weapons lose their broad synergy with martial arts, tonfa is added to specific martial arts.
* Tonfa is added to specific martial arts.
* Melee system is reworked in a way that allows tonfa to utilize weaker strikes that count as unarmed and stronger ones that count as bashing.
* Tonfa doesn't change.

**Comment by Izicata** [comment]
>Just because the game has faults doesn't mean we will accept all shit that makes them worse.

Having tonfas not usable with martial arts is a fault. This PR fixes that fault.

>It's your job to present an argument for why will it improve the game, not our to tell you why not, at least until you present a valid case for your side.

Realism and consistency. As the tonfa is primarily a weapon designed for use with martial arts, it should be able to be used with martial arts, just like the other weapons which can be used with martial arts.

>Tonfa is "less unarmed" than brass knuckles. It won't support simple jabs as well as brass knuckles and requires different movement to utilize fully. Boxer won't utilize tonfa well, but could do fine with brass knuckles.

No they're not. Tonfas are primarily designed for use with martial arts; brass knuckles are primarily designed for use by people untrained in martial arts. Tonfas are _more_ of an unarmed weapon than brass knuckles are. I even posted a [youtube video](https://www.youtube.com/watch?v=35gzIUOzb60) showing how the movements are the same. When holding the tonfa by the side-handle with the long end pointed towards your elbow, it can augment almost all hand and arm strikes, including punches, elbow strikes, and knife-hand strikes (karate chops). Fast jabs are possible with a tonfa.

>Someone not trained in unarmed combat will utilize tonfa as a club.

Someone not trained in unarmed combat might try to use the a katar as a regular dagger. Does that mean we should give it two sets of stats, one with lower to-hit than the other?

>Tonfa's stats are for being used as a club, not as brass knuckles. For example, relatively high damage that requires momentum. Brass knuckles are harder than a tonfa, but hit weaker because they do not get that momentum.

The stats can very easily be changed. For example, lowering the base damage and replacing the rapid strike with a technique that has a bashing multiplier to represent the spinning strikes. If this objection had been raised earlier, I would have done so.

>Tonfa is closer to held rock than to brass knuckles.

This is just blatantly wrong. Did you even watch the video? It's quite short.

>Tonfa is added to specific martial arts.

>And, as I've already said, the tonfa can be used to augment almost all hand and arm strikes, so that would be basically all of the martial arts except Eskrima.

**Comment by Coolthulhu** [comment]
> Having tonfas not usable as an unarmed augmenting weapon is a fault.

It isn't. Not all martial arts utilize tonfa strikes.

> As the tonfa is primarily a weapon designed for use with martial arts, it should be able to be used with martial arts, just like the other weapons which can be used with martial arts.

Then add it to specific martial arts that it is designed for, without adding it to ones it isn't designed for, such as ones based on kicks or grappling.

> Someone not trained in unarmed combat might try to use the a katar as a regular dagger. Does that mean we should give it two sets of stats, one with lower to-hit than the other?

If both have a reason to be used then sure.
Tonfas being used as clubs is nothing new or unusual.

Removing the ability to use katar for all unarmed would also be a good idea. Otherwise we get situations where you can axe-kick and throw with katar.

Nerfing tonfa and making it usable as unarmed would also work.

> This is just blatantly wrong. Did you even watch the video? It's quite short.

I did. Not all of it, but enough to notice that those movements are quite different from unarmed punches. Trying a heavy boxing punch with one could easily end up with a broken wrist.

It certainly isn't designed with all martial arts in mind, especially if we consider the in-game definition of "martial art".


Tonfa could use the unarmed combat skill, but it shouldn't work with all martial arts.
Same for all sorts of punch daggers. But those are already in the game and unarmed tonfa isn't and it won't get in just on "but punch daggers also have same faults".

**Comment by DangerNoodle** [comment]
> Also NO_DISARM is overly redundant - why can't we just check for NO_UNWIELD?

Because `NO_UNWIELD` is not for unarmed weapons, it is for bionic weapons, handcuffs, cursed items, whatever else is explicitly supposed to be stuck to the player.

> Tonfa is closer to held rock than to brass knuckles.

I am not sure what your intended point here is, but it brings up the mental image of rock-based material arts.

**Comment by Zireael07** [comment]
>Then add it to specific martial arts that it is designed for, without adding it to ones it isn't designed for, such as ones based on kicks or grappling.

@Coolthulhu @Izicata  Let's take a look at the arts. + for those that might conceivably use tonfa, - for others (with reasoning)
Looking at the wiki, I suppose the martial arts are:
     Karate +
    Judo - (grappling)
    Aikido ?
    Tai Chi ?
    Taekwondo - (kicks)
    Capoeira - (kicks)
    Krav Maga ?
    Muay Thai ?
    Ninjutsu ?
    Zui Quan - (drunken boxing)

Brawling/Boxing - (at least not boxing, according to @Coolthulhu's comment above)

I am not listing the shaolin or venom styles as I have no idea how they work.

Wikipedia doesn't want to work so I only listed those I can figure out from CataDDA, comments above or my friends (one friend trained karate briefly, another trains taekwondo)
> Removing the ability to use katar for all unarmed would also be a good idea. Otherwise we get situations where you can axe-kick and throw with katar.
+1.

**Comment by Izicata** [comment]
I know for a fact that you can use the tonfa with taekwondo. Do I need to post a timestamped picture of my third degree black belt certificate?

**Comment by Coolthulhu** [comment]
No, but you'll need a video of someone using a tonfa to augment an axe-kick, side-kick or sweep-kick.

**Comment by DangerNoodle** [comment]
That sounds more like an issue with the way kicks are coded, not the tonfas. Fixing techniques that do not make sense for weapons should be the subject for a different pull request.

**Comment by Izicata** [comment]
>No, but you'll need a video of someone using a tonfa to augment an axe-kick, side-kick or sweep-kick.

Here's [one](https://www.youtube.com/watch?v=iRrymKogde4), [two](https://www.youtube.com/watch?v=GphAeqYYoiI) videos of side-kick and punch combo strikes, done in much less time than a single turn (6 seconds). This is applicable for all unarmed weapons.

**Comment by DangerNoodle** [comment]
Even then, this problem would be sidestepped entirely if we approached the anomaly as "we can kick with a hand-held weapon, therefore kicks need to be fixed" instead of "we can kick with a handheld weapon, therefore weapons need to be fixed."

**Comment by Izicata** [comment]
Maybe a technique flag that requires the user to be completely unarmed? I agree that doing a throw should require you to be completely unarmed; you can't be holding something else while you're throwing someone.

**Comment by Coolthulhu** [comment]
> Even then, this problem would be sidestepped entirely if we approached the anomaly as "we can kick with a hand-held weapon, therefore kicks need to be fixed" instead of "we can kick with a handheld weapon, therefore weapons need to be fixed."

This argument won't work here. It isn't the first time an addition is held up for making other problems worse.

> Here's one, two videos of side-kick and punch combination strikes, done in much less time than a single turn (6 seconds).

Not kick and punch, just the kick.
Otherwise you could argue that guns are unarmed weapons because you can shoot someone and then kick him.

> Maybe a technique flag that requires the user to be completely unarmed?

I'd rather have it addressed directly at the source of the problem:
By making unarmed weapons not work with martial arts until otherwise specified, then both sides buffed to make them useful separately.

Changing individual techniques would require good UI, otherwise it would be hard to tell why isn't a technique working.

**Comment by Izicata** [comment]
>I'd rather have it addressed directly at the source of the problem:
>By making unarmed weapons not work with martial arts until otherwise specified, then both sides buffed to make them useful separately.

Then code it yourself. This tonfa implementation is completely consistent with how unarmed weapons and martial arts currently work. If you want to entirely rework the martial arts system, go right ahead, but I'm fine with how it works right now.

**Comment by Coolthulhu** [comment]
> Then code it yourself.

Sure.
Until then, tonfa stays a weird club.

**Comment by Izicata** [comment]
I'm sure this martial arts rework will definitely happen someday. Maybe within the next decade?

**Comment by DangerNoodle** [comment]
For the time being, instead of the both of you arguing over the matter, perhaps simply save the tonfa usage changes for a separate pull request? This lays a usable foundation for future additions,namely the possibility of unarmed weapons that can be disarmed.

While the tonfa is potentially a good example of such, an example is not essential for this pull request, even less so if it becomes a cause for bickering.

If we had a genuine katar or pata, instead of a punch dagger, that would be a better way to showcase this combination of features, but even then it is subject to the "kick while armed" problem.

**Comment by Izicata** [comment]
>While the tonfa is potentially a good example of such, an example is not essential for this pull request, even less so if it becomes a cause for bickering.

I'm afraid it is. If code isn't used, it doesn't get added. See Coolthulu's comment on #17806.
>The conductive flag isn't used and so shouldn't be added in the PR.

If tonfas don't get changed, none of the other changes get used and as such can't be merged, except maybe the minor bugfixes.

**Comment by Coolthulhu** [comment]
> I'm sure this hypothetical martial arts rework will definitely happen someday. Maybe within the next decade?

Yeah, things look bleak for the unarmed tonfa.
Unless someone proposes something that will let it nicely integrate with the game, it will stay out there in the realm of badly though out ideas.

It doesn't need to be a full rework, but it will be more work than adding 2 flags and calling it a day.

**Comment by mugling** [comment]
> It's your job to present an argument for why will it improve the game, not our to tell you why not, at least until you present a valid case for your side.

This has come up a few times recently. In opening a PR you are expected to make your case for a change vs the status quo. You are trying to convince a developer - not the reverse.

> If tonfas don't get changed, none of the other changes get used and as such can't be merged

If you feel it cannot be merged then you should consider closing your PR.

**Comment by DangerNoodle** [comment]
> Unless someone proposes something that will let it nicely integrate with the game, it will stay out there in the realm of badly though out ideas.

This means that use of weapons in unarmed-only attacks will remain a stumbling block. Allowing specific attacks to ignore wielded weapon would provide the consistency fix, but what would make it a desirable feature is if the unarmed-only attacks were furthered fleshed out to make up for not using the direct damage. At a bare minimum, these attacks would likely be faster depending on how slow the current wielded weapon is, if they also use "no weapon" for the purpose of calculating attack speed.

What pushes this idea further into being useful on its own is if the material art info screen clarified which techniques do not use any weapon, which techniques only use unarmed weapons, and which techniques are for weapons only.

Then if the player was given some control over which class of technique is used, they could force their character to favor attacks based on the situation. Such as use of faster unarmed-only attacks when they want more attacks per turn, then use the armed techniques when more damage is desired.

**Comment by Izicata** [comment]
>This has come up a few times recently. In opening a PR you are expected to make your case for a change vs the status quo. You are trying to convince a developer - not the reverse.

I did make a case, in the very first post of the very first PR: "As the tonfa is primarily a weapon designed for use with martial arts, it should be able to be used with martial arts. When holding the tonfa by the side-handle with the long end pointed towards your elbow, it can augment almost all hand and arm strikes, including punches, elbow strikes, and knife-hand strikes (karate chops)." 

In short, realism and consistency.

This PR can be merged as soon as you stop letting the perfect be the enemy of the good. The martial arts rework isn't going to happen, so why let a hypothetical perfect rework prevent refinements of the current system? It's not like this PR is going to prevent said hypothetical perfect rework from happening.

**Comment by Coolthulhu** [comment]
> In short, realism and consistency.

It's not realistic to augment a kick by wielding a tonfa.
It's consistent in that it's equally as stupid as using a punch dagger to do the same. ie. in a way that it shouldn't be.

> This PR can be merged as soon as you stop letting the perfect be the enemy of the good.

No, it won't be merged until you stop making horrible the friend of bad.
At the moment it exacerbates existing problems, just for "realism" that allows striking with tonfas by kicking. One tiny step forward, two steps back.

You could limit it to the cases where it makes sense, like martial arts that would sanely work with tonfa and don't use more than one kick technique.
You could invent some other way of making it work with martial arts.
You can propose a list of alternative ideas and we will tell you which ones don't suck.

It's not an "on the fence" situation that can be out-argued with clever semantics or "but katar".
Tonfa isn't getting the flag that allows it to work with martial arts that rely mostly on kicking and grappling.

**Comment by DangerNoodle** [comment]
In that case, supposing that we are forced to focus on the perquisites and foundation, then I will outline my proposals here.

In this case we have a end goal of implementing unarmed weapons that can be disarmed. The anomaly of techniques that can use unarmed weapons (but logically should not) is identified as an obstacle to this.

Solutions to this range from simple to complex. The solution I have proposed is to allow specific techniques to ignore the weapon wielded and count as an attack with no weapon. Two obstacles to this have been identified. First, a pre-existing lack of information regarding which techniques make use of weapons. Second, the need for further benefits to actually making this distinction.

The solution to the first obstacle is self-evident and would be useful to players even on its own merit, by giving some way to tell players whether a given technique counts as unarmed, armed, or both. The second obstacle would be resolved by making the unarmed-only techniques actually worth using even with a weapon wielded. While the increased attack speed helps, fleshing out the different techniques would help, and also have the benefit of making the different martial arts more distinct, giving different niches for them.

And lastly, giving players further control over their use of techniques potentially complements this. It is useful even independent of these ideas, and would be eevn more useful with them, but (my opinion) not a strict prerequisite to this series of pull requests.

Thus, we have a possible workflow:
1. Expand the material arts info screen to clarify which techniques are unarmed, and which are not. This is the most immediately useful and potentially the easiest to implement.
2. Flesh out techniques in various ways. Use of knockback, stunning, prone, different attack speed multipliers and such could be codified to specific standards. Kicks might tend towards a speed reduction (or affect post-attack recovery only) while being the main source of knockback, throws could be standardized around flooring enemies, etc.
3. Implement a flag or other property flagging a technique as unarmed-only, and add code to make these techniques treat fists as the base weapon regardless of weapon wielded. This fixes our end goal's immediate prerequisite, and is supported by previous additions.
4. This pull request, allowing for unarmed weapons that can be disarmed.
?. Giving players a menu to toggle unarmed-only, unarmed-weapon, armed-weapon, and other classifications of technique. Compliments all of these additions, but exactly placement is uncertain.

**Comment by Izicata** [comment]
>It's not realistic to augment a kick by wielding a tonfa.

one attack is an abstraction for multiple strikes & video of combination attack, continued assertions that you can't do that, rinse and repeat

>It's consistent in that it's equally as stupid as using a punch dagger to do the same. ie. in a way that it shouldn't be.

So no new weapons can gain the unarmed_weapon flag until the martial arts system is reworked, which you seem unwilling to do yourself. Okay.

**Comment by Izicata** [comment]
>2. Flesh out techniques in various ways. Use of knockback, stunning, prone, different attack speed multipliers and such could be codified to specific standards. Kicks might tend towards a speed reduction (or affect post-attack recovery only) while being the main source of knockback, throws could be standardized around flooring enemies, etc.

While we're making wishes, kick damage and speed should be effected by the weight and material of the shoes you're wearing. Maybe even bonuses for specific makes of shoes, like steel-toed boots or survivor-made shoes with pointy bits on the toe and heel.

**Comment by DangerNoodle** [comment]
That is interesting but unlikely to occur. At least altering and balancing techniques will be primarily JSON work.

**Comment by Izicata** [comment]
If kicks aren't going to be boosted by hand weapons because they're a foot technique, they should be boosted by foot weapons. Which shoes are, just as much as brass knuckles or boxing gloves are hand weapons.

**Comment by DangerNoodle** [comment]
Useful, but not mandatory to fleshing out technique. If it is, remember this further complicates the series of changes that compliment the change you have insisted on.

This is giving me flashbacks to Mining Mod.

**Comment by Coolthulhu** [comment]
> one attack is an abstraction for multiple strikes

Until stated otherwise, it isn't. It's the 6 seconds that are an abstraction here.

> So no new weapons can gain the unarmed_weapon flag until the martial arts system is reworked, which you seem unwilling to do yourself. Okay.

Only those which use """"realism"""" as an argument that allow striking with a tonfa to be a kick.
You could try something like "we need a weapon that does x and tonfa could do that".
But you went with realism that you couldn't defend without pushing boundaries of both in-game and out-of-game ideas and very subjective "consistency" of what defines an unarmed weapon.

And then you added passive-aggressive attitude to make sure that you won't get any leniency in standards.

How Not To Talk To People Who Filter Bad PRs: The PR.

If it wasn't me reviewing, this would most likely be closed long ago. I'm still trying to give you the benefit of doubt here.

**Comment by Izicata** [comment]
>Second PR, second reviewer, same issues. Please refrain in future from duplicate PR's (this isn't the first time either).

I closed the previous PR myself because it had merge conflicts and was easier to recreate than to update. Hardly a duplicate.

**Comment by mugling** [comment]
There was a comment immediately after suggesting not to add tonfas - which you just ignored. Asides, you've done it before on other occasions. So yes, exactly a duplicate.

**Comment by Izicata** [comment]
I replied with two paragraphs about why tonfas should have unarmed_allowed. That's not what ignoring means.

**Comment by DangerNoodle** [comment]
If @mugling would cite what this suggestion was, which respect to why tonfas should not be added?

If the objection is the "kicks while wielded" anomaly, I have already outlined a plan to salvage this idea. All three of you are getting caught up in an argument that serves no purpose, instead of addressing the mechanisms that would potentially turn this idea into part of a more useful series of changes.

**Comment by mugling** [comment]
The first review said no. You should have continued discussion in that thread as opposed to trying for a new PR and asking for a second review.

**Comment by Izicata** [comment]
Here:
>I'm still not sure if unarmed on tonfas is a good thing.
We have those weird unarmed attacks which allow kicks to benefit from brass knuckles and all.
It should probably just be allowed for specific martial arts.

The same objection was posted by the same reviewer earlier up the thread in #18543. I had already posted a rebuttal to it and got no response for 15 days; the new comment did not address any of the points in either the original rebuttal or the second rebuttal, it was just a duplicate of the original objection that ignored my responses. That's stonewalling, not discussion.

**Comment by DangerNoodle** [comment]
And that is the exact objection that I furthermore presented an alternative solution to. Could you at least address my suggested series of prerequisite changes?

**Comment by Coolthulhu** [comment]
@DangerNoodle This could work for the kick part, not grappling. It would also require a good UI that adequately communicates what will and what will not work with a weapon.

It would be enough to add tonfa to martial arts that use kicks, but not enough to add it as generic martial arts item.

**Comment by Izicata** [comment]
Grappling techniques could get a flag that requires the user to be completely unarmed, that is, wielding a none? Looking in MARTIALART_JSON.md, apparently such a flag used to exist: `strict_unarmed`.

**Comment by DangerNoodle** [comment]
Regarding the first idea, grappling may or may not work with the "unarmed only" flag. It might be reasonable to assume that medieval swordsmanship grabs would allow using the weapon as leverage, but not unarmed styles. But that could just as easily be resolved by making medieval swordsmanship use a different grappling technique that counts as armed, not unarmed.

Regarding the second point, I did propose a rough step in the right direction, by making the martial art info screen divide the technique list based on whether they allow a weapon.

Regarding the third, that might be an alternative way to handle it, yes.

**Comment by Izicata** [comment]
Medieval swordsmanship already uses a different `melee_allowed` grab technique.

**Comment by DangerNoodle** [comment]
Ah, that covers that issue then.
